README.md
## muReq - Multiple Request  
[![Build Status](https://travis-ci.org/DarkScorpion/muReq.svg?branch=m)](https://travis-ci.org/DarkScorpion/muReq)
[![LICENSE](https://img.shields.io/badge/license-MIT-blue.svg)](https://travis-ci.org/DarkScorpion/muReq)
[![Downloads](https://img.shields.io/npm/dt/mureq.svg)](https://travis-ci.org/DarkScorpion/muReq)
[![Downloads in Month](https://img.shields.io/npm/dm/mureq.svg)](https://travis-ci.org/DarkScorpion/muReq)


Node.js module, for multiple and single requests with using Promises. It takes an array or object with the addreses of the request. Returns Promise with an array or object containing the result. Flexible customization down to the individual fields of the request.  

### Install
```
npm install mureq
```

### Examples
#### Start
```js
var muReqClass = require('mureq');
var muReq = new muReqClass(); //default settings
var muReqCustom = new muReqClass( {errCoef: 0.9} ); //custom settings
```
or use one-line init script (can use settings object)
```js
var muReq = require('mureq/init')();
```

#### Multi request
```js
var url = 'http://test.zz/';

//Multi request (array)
muReq.get( [ url+'1', url+'2', url+'3' ] )
  .then( (result) => {
    console.log(result) //=> [ 'get 1', 'get 2', 'get 3' ]
  })

//Multi request (object)
muReq.get( { a1: url+'1', a2: url+'2', a3: url+'3' } )
  .then( (result) => {
    console.log(result) //=> { a1: 'get 1', a2: 'get 2', a3: 'get 3' }
  })
```

#### Single request
```js
muReq.get('http://test.zz/')
  .then( (result) => {
    console.log(result) //=> 'ok'
  })
  .catch( (err) => {
    console.log(err)
  });
```

#### Specify the multi request method (get, post, put, delete)
```js
var urls = [ url+'1', url+'2', url+'3' ];

muReq.post(urls)
  .then( (result) => {
    console.log(result) //=> [ 'post 1', 'post 2', 'post 3' ]
  })

muReq.put(urls)
  .then( (result) => {
    console.log(result) //=> [ 'put 1', 'put 2', 'put 3' ]
  })
```

#### Customization requests
For custom request, use object with field of module [request](https://www.npmjs.com/package/request)
```js
//Custom metods
var reqObj = {
  a1: { url: url+'1', method: 'post' },
  a2: url+'2',
  a3: { url: url+'3', method: 'put' }
}
muReq.get(reqObj)
  .then( (result) => {
    console.log(result) //=> { a1: 'post 1', a2: 'get 2', a3: 'put 3' }
  })
```

#### Cloning configuration request fields
If you want to make requests with certain fields at different urls, use the static method prepareData
```js
var url = 'http://test.zz/';
var urls = [ url+1, url+2, url+3 ];
var options = {
  method: 'POST',
  headers: {
    'User-Agent': 'request'
  }
};

var result = muReq.prepareData(urls, options);

console.log(result) 
/* => [
  { url: 'http://test.zz/1', { method: 'POST', headers: { 'User-Agent': 'request' } } },
  { url: 'http://test.zz/2', { method: 'POST', headers: { 'User-Agent': 'request' } } },
  { url: 'http://test.zz/3', { method: 'POST', headers: { 'User-Agent': 'request' } } }
]
*/
```

#### Settings
You can customizate sittengs for you tasks.
```js
{ //Field of settings and their default value
  errCoef: 0.2, // if errors more 20%, promise return error;
  checkStatusCode: true, // 404 and other status, considered an error
  promise: Promise // native Node.js promise
}
//Customizate example and check settings
muReq.setSettings( {errCoef: 0.8} );
console.log( muReq.getSettings() ); //=> { errCoef: 0.8, checkStatusCode: true }
```

#### Use other promise library
You can specify a different Promis library. Bluebird package (v 3.4.7) tested and works.
```js
var muReq = new muReqClass( { promise: require('bluebird') } );
```
or
```js
muReq.setSettings( { promise: require('bluebird') } );
```

#### Errors
Function return body of response, if not errors and status code 200. 
```js
var reqData = [
  url+'1',
  url+'/route404',
  'notUrl'
];
muReq.get(reqData)
  .then( (result) => {
    console.log(result) //=> [ 'get 1', Error('Status code, is not 200'), Error('Invalid URI "notUrl"') ]
  })
```
If not need check status code, change settings.
```js
muReq.setSettings( {checkStatusCode: false} );
muReq.get(reqData)
  .then( (result) => {
    console.log(result) //=> [ 'get 1', 'route 404 data', Error('Invalid URI "notUrl"') ]
  })

```


CHANGELOG.md
## Change Log

### v1.2.4
- add bitbacket pipelines

### v1.2.1
- add script for one-line create object of class.
- update test.

### v1.2.0
- Add new field in settings: promise
- Tested compatibility with "bluebird" package.
- Update readme (badges).

### v1.1.1
- Update readme.
- Update tests.

### v1.1.0
- Add static method prepareData, for cloning request options for different urls.

### v1.0.0
- muReq became a class and need to initialize using, the "new".
- muReq start use params: errCoef, checkStatusCode.

### v0.9.0
- Single method, for multiple request.


test/errors.js

var errors = require('../lib/errors.js')

errors.notUrl = new Error('Invalid URI "notUrl"');

module.exports = errors;


test/server.js
'use strict';

var express = require('express');
var app = express();

app.get('/', (req, res) => {
  res.send('ok');
});

//Query routes
app.get('/query', (req, res) => {
  var q = req.query;
  var p1 = q.p1 || '???';
  var p2 = q.p2 || '???';
  var result = p1+' '+p2;
  res.send(result);
});

//Param routes
app.get('/echo/:p1', (req, res) => {
  var p1 = req.params.p1;
  res.send(p1);
});

//Types routes
app.all('/types', (req, res) => {
  res.send(req.method + ' ok');
});

//Route404
app.get('/route404', (req, res) => {
  res.status(404).send('404');
});

//Other routes
app.all('*', (req, res) => {
  console.log('Warning: strange route: %s %s', req.path, req.method);
  res.status(404).send('=?=');
});

module.exports = app;


test/t4 error.js
'use strict';
var assert = require('assert');

var app = require('./server.js');
var muReq = require('../init.js')( {errCoef: 0.8} ); //test fast init script with params

const PORT = 3004;
const SERVER = app.listen(PORT);
const BASE_URL = 'http://localhost:'+PORT;

const ERRORS = require('./errors.js');

describe('Error in requests', () => {
  describe('Multiple', () => {
    it('Array', (done) => {
      var input = [
        BASE_URL+'/',
        BASE_URL+'/types',
        BASE_URL+'/route404',
        'notUrl'
      ];
      var output = [
        'ok',
        'GET ok',
        ERRORS.not200,
        ERRORS.notUrl
      ];

      testMuReq(input, output, done);
    });

    it('Object', (done) => {
      var input = {
        a: BASE_URL+'/',
        b: BASE_URL+'/types',
        c: BASE_URL+'/route404',
        d: 'notUrl'
      };
      var output = {
        a: 'ok',
        b: 'GET ok',
        c: ERRORS.not200,
        d: ERRORS.notUrl
      };

      testMuReq(input, output, done);
    });
  });

  describe('Single', () => {
    it('Standart', (done) => {
      testErr('notUrl', ERRORS.notUrl, done);
    });
  });

  after( () => {
    SERVER.close();
  });

});

function testMuReq(input, output, done) {
  muReq.request(input)
    .then( (result) => {
      assert.deepEqual(output, result);
      done();
    })
    .catch( (err) => {
      assert.fail();
      done();
    });
}

function testErr(input, output, done) {
  muReq.request(input)
    .then( (result) => {
      assert.fail();
      done();
    })
    .catch( (err) => {
      assert.deepEqual(output, err);
      done();
    });
}



lib/multi-request.js
'use strict';
//Copyright (c) 2016 Александр Смит (https://github.com/DarkScorpion)
var request = require('request');
var typeOf = require('./typeOf.js');

const DEF_SETTINGS = {
  errCoef: 0.2,
  checkStatusCode: true,
  promise: Promise
};

const ERROR = require('./errors.js');

class muReq {
  constructor(opt) {
    this._settings = Object.assign({}, DEF_SETTINGS, opt);
  }

  setSettings(opt) {
    this._settings = Object.assign({}, this._settings, opt);
  }

  getSettings() {
    return Object.assign({}, this._settings);
  }

  get(reqData) {
    return this.request(reqData, 'get');
  }

  post(reqData) {
    return this.request(reqData, 'post');
  }

  put(reqData) {
    return this.request(reqData, 'put');
  }

  delete(reqData) {
    return this.request(reqData, 'delete');
  }

  request(reqData, reqType) {
    var rdType = typeOf(reqData);
    if (rdType === 'string') {
      return this._singleRequest(reqData, reqType);
    }
    if (rdType === 'array' || rdType === 'object') {
      return this._multiRequest(reqData, reqType);
    }

    throw ERROR.mustStrArrObj;
  }

  static prepareData(reqData, reqOptions) {
    var rdType = typeOf(reqData);

    if(rdType !== 'array' && rdType !== 'object') {
      return ERROR.mustArrObj;
    }
    if( typeOf(reqOptions) !== 'object' ) {
      return ERROR.mustObj;
    }

    var result = (rdType === 'array') ? new Array(reqData.length) : {};

    for(let key in reqData) {
      var value = reqData[key];
      var pushData = ( typeOf(value) === 'object' ) ? value : {url: value};

      result[key] = Object.assign({}, reqOptions, pushData);
    }

    return result;
  }

  _multiRequest(reqData, reqType) {
    return new this._settings.promise( (resolve, reject) => {
      var counter = 0;
      var errCounter = 0;

      var rdType = typeOf(reqData);
      var reqMethod = reqType || 'GET';
      var notCheck = !this._settings.checkStatusCode;

      var length = rdType === 'array' ? reqData.length : Object.keys(reqData).length;
      var result = rdType === 'array' ? new Array(length) : {};

      for(let key in reqData) {
        var value = reqData[key];
        var options = ( typeOf(value) === 'object' ) ? value : {url: value, method: reqMethod};
        request(options, (err, res, body) => {
          if ( !err && (notCheck || res.statusCode === 200) ) {
            result[key] = body;
            counter++;
          } else {
            result[key] = err || ERROR.not200;
            errCounter++;
            counter++;
          }

          if (counter === length) {
            var isManyErr = errCounter/length > this._settings.errCoef;
            if (isManyErr) {
              reject(ERROR.manyErr);
            } else {
              resolve(result);
            }
          }

        });
      }

    });
  }

  _singleRequest(reqUrl, reqType) {
    return new this._settings.promise( (resolve, reject) => {
      var options = {
        url: reqUrl,
        method: reqType || 'GET'
      };
      
      var notCheck = !this._settings.checkStatusCode;
      request(options, (err, res, body) => {
        if ( !err && (notCheck || res.statusCode === 200) ) {
          resolve(body);
        } else {
          reject(err || ERROR.not200);
        }
      });

    });
  }
}

module.exports = muReq; 

